Add a `frequency` parameter to `TriggerWandbSyncLightningCallback`

CedricLeon opened the pull request:
In order to synchronize every N epoch instead of every epoch (by default), I added a frequency parameter to the Lightning callback.

I did not write any test for that feature, mainly because I am not sure how you proceed (but also because of lazyness 😅)
Tell me if you want me to add a test for that.

CedricLeon committed:
Add a `frequency` parameter to `TriggerWandbSyncLightningCallback`

klieret commented on src/wandb_osh/lightning_hooks.py:
```diff
@@ -41,6 +43,7 @@ def __init__(
         """
         super().__init__()
         self._hook = TriggerWandbSyncHook(communication_dir=communication_dir)
+        self.frequency = frequency
```
could you make this private? i.e., `self._frequency`

CedricLeon replied:
Yes, good idea

klieret commented on src/wandb_osh/lightning_hooks.py:
```diff
@@ -24,11 +24,13 @@ class TriggerWandbSyncLightningCallback(pl.Callback):
     def __init__(
         self,
         communication_dir: PathLike = _comm_default_dir,
+        frequency: int = 1,
```
I wonder if there's a better name for this. Because this setting is basically `1/frequency`, right? (the larger this value is, the _less_ often we sync). So perhaps something like `sync_every_n_epochs`?

CedricLeon replied:
I started with such a name but thought that frequency was shorter and still self-explanatory.
I will make the change to `sync_every_n_epochs`, it is more accurate

klieret commented on src/wandb_osh/lightning_hooks.py:
```diff
@@ -49,4 +52,6 @@ def on_validation_epoch_end(
             return
-        self._hook()
+        # Call the hook every {frequency} epoch
+        if trainer.current_epoch % self.frequency == 0:
```
I think this line should be implemented in `TriggerWandbSyncHook` (i.e., we pass the `frequency` argument there during `__init__` and then just call `self._hook()` here. This way, all derived classes can use this and not just the one for pytorch lightning (and then we add the same passing of the `frequency` argument to `TriggerWandbSyncRayHook`).

CedricLeon replied:
That's better indeed, I will make the change

CedricLeon committed:
Renaming `frequency` in `sync_every_n_epochs`

CedricLeon committed:
Add a `sync_every_n_epochs` parameter to `TriggerWandbSyncHook()` initialization
Now the calling `hook()` must be done by giving information about the current epoch

CedricLeon committed:
Updating the Lightning Hook, with the new `TriggerWandbSyncHook` initialization and call

CedricLeon committed:
Updating the Ray Hook, with the new `TriggerWandbSyncHook` initialization and call

CedricLeon commented:
I made the changes. Still without testing 😇
I 100% confident it works for the Lightning hook (`TriggerWandbSyncLightningCallback`), because i use such a setup in my work. However, I never used ray before and can't think of an easy way to test it. Can you have a look and tell me if I implemented something stupid?
Also about #83, I think it can definitely help but won't solve the problem.

klieret commented:
Oh you're 100% right, I should have thought about that as well. A totally different avenue of solution (that might also help with syncing hyperparameters of runs that get killed because of #100) would be that the `wandb-osh` syncer tool remembers which runs were syncing and if it doesn't get triggered for an hour, it runs one final sync call. Of course that's also just another silly workaround.

CedricLeon commented:
That would be a pretty robust workaround indeed.
Should I add the `on_test_epoch_end()` fix I mentionned and we close this PR?
Or should I just add a Warning about this problem in the doc, and we fix it in another PR?

I dug into the problem I raised in #100 and this is deeper that just not syncing the config. i will detail what I found so far in the Issue.
Anyway, it would be handy for me to have the possibility to reduce the number of Triggers to investigate when the problem happens.

CedricLeon commented:
@klieret should I push this temporary fix?
Or do you want to solve it in this PR?

CedricLeon committed:
Add the `on_test_epoch_end()` Lightning hook to force the synchronization on the end of the last epoch

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CedricLeon committed:
Correct stupid typo in lightning_hooks.py

klieret commented:
Could you give me push permission on the branch (I made some fixes and added some tests to this). You [only have to click a checkbox on this PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), I think

CedricLeon commented:
> Could you give me push permission on the branch (I made some fixes and added some tests to this). You [only have to click a checkbox on this PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), I think

The box is checked, it seems to be checked by default btw

CedricLeon commented:
Hey @klieret, I found this old PR. Is that anything you're still working on? It seems your package has evolved quite significantly; maybe we should just close that one.
I'll leave it up to you

CedricLeon committed:
Merge branch 'klieret:main' into main
